New Tutorial: Parameter tuning and optimization - Evaluating nuclei segmentation with Galaxy (#5988)
galaxyproject/training-material: topics/imaging/tutorials/parameter-tuning/tutorial.md

  - How can we choose the best filters for preprocessing images for nuclei segmentation?
objectives:
  - Understand the importance of parameter tuning in bioimage analysis for achieving accurate results
  - Learn how to perform parameter tuning for segmentation in Galaxy
key_points:
  - Parameter tuning helps achieve reliable and repeatable results
time_estimation: "1H"
follow_up_training:
  -
    type: "internal"
    topic_name: imaging
    tutorials:
      - hela-screen-analysis
contributions:
  authorship:
    - rmassei
  funding:
    - nfdi4bioimage
    - dfg
tags:
  - bioimaging

---

Parameter tuning is super important in bioimaging. 
When you're doing image analysis — like segmentation, quantification, or feature extraction — the settings you choose make a big difference in how accurate your results are. 
But biological images can vary strongly in terms of quality, contrast, and structure. 
So you can't just use the same settings for every image. 

That's where parameter tuning comes in. 
By adjusting these settings carefully, researchers can make sure their tools are capturing the right biological info without adding noise or messing up the images. 

> This helps make sure their work is repeatable and reliable, which is key getting accurate scientific results. 

Reviewer: ```suggestion
This helps make sure their work is repeatable and reliable, which is key for getting accurate scientific results. 
```

>     input of {% icon tool %} **6: Filter 2-D image**. 
>     - Connect the output of {% icon param-file %} *6: Filter 2-D image** to the "Input Image"
>     input of {% icon tool %} ** 7: Perform histogram equalization **. 
>     - Connect the output of {% icon param-file %} **7: Perform histogram equalization** to the "Input Image"
>     input of {% icon tool %} **8: Threshold image**. 
>     - Connect the output of {% icon param-file %} **8: Threshold image** to the "Segmented images"
>     input of {% icon tool %} **9: Compute image segmentation and object detection performance measures**. 
>     - Connect the output of {% icon param-file %} **2: Input Ground Truth** to the "Ground truth images"
>     input of {% icon tool %} **9: Compute image segmentation and object detection performance measures**. 
>     - Connect the output of {% icon param-file %} **9: Compute image segmentation and object detection performance measures** to the "Collection of files to collapse into single dataset"
>     input of {% icon tool %} **10: Collapse Collection**. 
>     - Connect the output of {% icon param-file %} **10: Collapse Collection** to the "and"

> >     input of {% icon tool %} **11: Paste**.

Author: Nice catch!

>     - Connect the output of {% icon param-file %} **3: Sigma values to test** to the "Paste"
>     input of {% icon tool %} **11: Paste**.
{: .hands_on}

> This is how the worflow should look like!

Reviewer: ```suggestion
This is how the workflow should look like!
```

![img.png](../../images/parameter-tuning/workflow_parameter_tuning.png)

In this workflow, the {% tool [Parse parameter value](param_value_from_file) %} tool is passing *sigma* values one by one to the workflow downstream. In this way, a collection
of datasets is created for each sigma value and each one is processed individually.
Such strategy can be adapted for any parameter of the workflow and, in general, for any tool.

> Coming back to the results, here below we can see a potential output of the analysis by running four different sigmas (0.5, 5, 10, 50)

Reviewer: ```suggestion
Coming back to the results, below we can see a potential output of the analysis by running four different sigmas (0.5, 5, 10, 50).
```

![img_1.png](../../images/parameter-tuning/table_results.png)

In this example, results suggest that as the sigma value increases,
the Dice coefficient, which measures the overlap between the predicted segmentation and
the ground truth, decreases. Lower sigma values result in segmentations 
that more closely match the ground truth, with the highest Dice coefficient of 
0.91 achieved at the lowest sigma value of 0.5. 

The Hausdorff 
distance, a measure of the maximum distance between two sets, significantly 
increases with higher sigma values, from 17.8 to 203. This increase shows that higher sigma values lead to segmentations 
that are not only less accurate but also potentially include large outliers or
misclassifications.

The influence of sigma on merge operations does not show a strong relationship which might indicate that 
there's an optimal range for sigma 
that balances detail preservation and noise reduction. 

> 

Reviewer 2: So the conclusion is that `sigma=0.5` is the optimal value? If so, and you have the time to do the work, I think it would be advisable to use a different image (because `sigma=0.5` is the lowest tested value here).

Preferably, a noisy image, because for such images I would expect that larger sigma will be advantageous. For example, there is the "GFP-GOWT1 mouse stem cells" training dataset on https://celltrackingchallenge.net/2d-datasets that also has 4 ground truth images.

Author: Makes sense, I will change it!

>     input of {% icon tool %} **6: Filter 2-D image**. 
>     - Connect the output of {% icon param-file %} **2: Dataset to perform segmentation** to the "Input Image"
>     input of {% icon tool %} **6: Filter 2-D image**. 
>     - Connect the output of {% icon param-file %} **6: Filter 2-D image** to the "Input Image"
>     input of {% icon tool %} **7: Perform histogram equalization**. 
>     - Connect the output of {% icon param-file %} **7: Perform histogram equalization** to the "Input Image"
>     input of {% icon tool %} **8: Threshold image**. 
>     - Connect the output of {% icon param-file %} **8: Threshold image** to the "Segmented images"
>     input of {% icon tool %} **9: Compute image segmentation and object detection performance measures**. 
>     - Connect the output of {% icon param-file %} **2: Input Ground Truth** to the "Ground truth images"
>     input of {% icon tool %} **9: Compute image segmentation and object detection performance measures**. 
>     - Connect the output of {% icon param-file %} **9: Compute image segmentation and object detection performance measures** to the "Collection of files to collapse into single dataset"

> >     input of {% icon tool %} **10: Collapse Collection**.

Reviewer 2: The formatting of the inputs is somewhat inconsistent here: There is `{% icon param-file %} *"File to split"* input` for the first bullet point (so icon + italic), which is nice I think, but then it is just like `"Object ID" input` for all the other bullet points below (no icon, not italic).

Author: Fixed!

    - dfg
tags:
  - bioimaging
---

Parameter tuning is super important in image analysis. 
When you're doing image analysis — like segmentation, quantification, or feature extraction — the settings you choose make a big difference in how accurate your results are. 
But biological images can vary strongly in terms of quality, contrast, and structure. 
So you can't just use the same settings for every image. 

That's where parameter tuning comes in. 
By adjusting these settings carefully, researchers can make sure their tools are capturing the right biological info without adding noise or messing up the images. 
This helps make sure their work is repeatable and reliable, which is key for getting accurate scientific results. 
Basically, it's all about finding the perfect balance to get the most useful and accurate data possible.

In this tutorial, we will show how to perform parameter tuning for segmentation. In particular, we will fine-tune a parameter of Gaussian filters for nuclei segmentation by
comparing ground-truth segmentation with the ones processed in Galaxy. As a dataset,
we will use GFP-GOWT1 mouse stem cell images, which were acquired from Bártová et al. (2011).

> The test dataset contains a collection of images plus their ground truth. The images are

Reviewer 2: The test dataset does not contain ground truth images, only the training dataset does:
```suggestion
The training dataset contains a collection of images plus their ground truth (Maška et al. 2014). The images are
```
where:
> M. Maška et al., “A benchmark for comparison of cell tracking algorithms,” Bioinformatics, vol. 30, no. 11, pp. 1609–1617, 2014.

This is where the ground truth data is from. People often also use this citation as the only citation of the dataset (so also for the nuclei images) but it is even better to also cite Bártová of course.

It would be cool if you could add this and Bártová et al. (2011). as proper citations with the `{% cite %}` directive :)

Author: Fixed, the bib in the tutorial.bib had also some typos, now everything should be clean

> publicly available at the [Cell Tracking Challenge](https://celltrackingchallenge.net/2d-datasets/) repository.

Reviewer 2: ```suggestion
publicly available at the [ISBI Cell Tracking Challenge](https://celltrackingchallenge.net/2d-datasets/) repository.
```

This helps make sure their work is repeatable and reliable, which is key for getting accurate scientific results. 
Basically, it's all about finding the perfect balance to get the most useful and accurate data possible.

In this tutorial, we will show how to perform parameter tuning for segmentation. In particular, we will fine-tune a parameter of Gaussian filters for nuclei segmentation by

> comparing ground-truth segmentation with the ones processed in Galaxy. As a dataset,

Reviewer 2: ```suggestion
quantitatively comparing ground-truth segmentation with the ones processed in Galaxy. As a dataset,
```

>    - **Ground truth is uniquely labeled**: ***No***
>    - Performance measure(s):

> >      - ***Region-based / DICE***

Reviewer 2: Dice is a proper name, not an abbreviation :) https://en.wikipedia.org/wiki/Dice-Sørensen_coefficient
```suggestion
>      - ***Region-based / Dice***
```

>     - Connect the output of {% icon param-file %} **2: Input Ground Truth** to the "Ground truth images"
>     input of {% icon tool %} **9: Compute image segmentation and object detection performance measures**. 
>     - Connect the output of {% icon param-file %} **9: Compute image segmentation and object detection performance measures** to the "Collection of files to collapse into single dataset"
>     input of {% icon tool %} **10: Collapse Collection**.
{: .hands_on}

This is how the workflow should look like!

![workflow_overview.png](../../images/parameter-tuning/workflow_overview.png)

In this workflow, the {% tool [Parse parameter value](param_value_from_file) %} tool is passing *sigma* values one by one to the workflow downstream. In this way, a collection
of datasets is created for each sigma value and each one is processed individually.
Such strategy can be adapted for any parameter of the workflow and, in general, for any tool.

> Coming back to the results, here the results from segmentation with four different sigmas...

Reviewer 2: ```suggestion
Coming back to the results, here are the results from segmentation with four different sigmas...
```

![sigma_comparisons.png](../../images/parameter-tuning/sigma_comparisons.png){: width="50%"}

...and below we can see the output of the analysis:

![output_workflow.png](../../images/parameter-tuning/output_workflow.png)

We are comparing three different performance measures region-based segmentation [Dice](https://en.wikipedia.org/wiki/Dice-S%C3%B8rensen_coefficient), 
[Jaccard](https://en.wikipedia.org/wiki/Jaccard_index) and [Rand](https://en.wikipedia.org/wiki/Rand_index).
All three indexes measure the agreement of the predicted segmentation image foreground and
the ground truth image foreground

> Overall, results suggest that as the *sigma* value increases,
> the coefficient values decreases. Lower sigma values result in segmentations 
> that more closely match the ground truth, with the highest Dice coefficient of 
> 0.74 achieved at the lowest sigma value of 5.

Reviewer 2: ```suggestion
Overall, the results suggest that as the *sigma* value increases too much,
the coefficient values decrease. Lower *sigma* values result in segmentations 
that more closely match the ground truth, with the highest Dice coefficient of 
0.74 achieved at a *sigma* value of 5.
```
